group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -0,0 +1,959 @@
+        CLOSED,
+
+        /**
+         * The coordinator has failed.
+         */
+        FAILED
+    }
+
+    /**
+     * CoordinatorContext holds all the metadata around a coordinator state machine.
+     */
+    class CoordinatorContext {
+        /**
+         * The topic partition backing the coordinator.
+         */
+        final TopicPartition tp;
+
+        /**
+         * The snapshot registry backing the coordinator.
+         */
+        final SnapshotRegistry snapshotRegistry;
+
+        /**
+         * The deferred event queue used to park events waiting
+         * on records to be committed.
+         */
+        final DeferredEventQueue deferredEventQueue;
+
+        /**
+         * The current state.
+         */
+        volatile CoordinatorState state;
+
+        /**
+         * The actual state machine.
+         */
+        volatile S coordinator;
+
+        /**
+         * The current epoch of the coordinator. This represents
+         * the epoch of the partition leader.
+         */
+        volatile int epoch;
+
+        /**
+         * The last offset written to the partition.
+         */
+        volatile long lastWrittenOffset;
+
+        /**
+         * The last offset committed. This represents the high
+         * watermark of the partition.
+         */
+        volatile long lastCommittedOffset;
+
+        /**
+         * Constructor.
+         *
+         * @param tp The topic partition of the coordinator.
+         */
+        private CoordinatorContext(
+            TopicPartition tp
+        ) {
+            this.tp = tp;
+            this.snapshotRegistry = new SnapshotRegistry(logContext);
+            this.deferredEventQueue = new DeferredEventQueue(logContext);
+            this.state = CoordinatorState.INITIAL;
+            this.epoch = -1;
+            this.lastWrittenOffset = 0L;
+            this.lastCommittedOffset = 0L;
+        }
+
+        /**
+         * Updates the last written offset. This also create a new snapshot
+         * in the snapshot registry.
+         *
+         * @param offset The new last written offset.
+         */
+        private void updateLastWrittenOffset(
+            long offset
+        ) {
+            if (offset <= lastWrittenOffset) {
+                throw new IllegalStateException("New committed offset " + offset + " of " + tp +
+                    " must be larger than " + lastWrittenOffset + ".");
+            }
+
+            log.debug("Update last written offset of {} to {}.", tp, offset);
+            lastWrittenOffset = offset;
+            snapshotRegistry.getOrCreateSnapshot(offset);
+        }
+
+        /**
+         * Reverts the last written offset. This also reverts the snapshot
+         * registry to this offset. All the changes applied after the offset
+         * are lost.
+         *
+         * @param offset The offset to revert to.
+         */
+        private void revertLastWrittenOffset(
+            long offset
+        ) {
```

Comment:
nit: do we want a check that is the opposite of the updateLastWrittenOffset (ie, offset can not be greater than last written offset?)